underscore.js

```diff
@@ -277,56 +277,33 @@
     return _.find(obj, _.matcher(attrs));
   };
 
-  // Return the maximum element (or element-based computation).
-  _.max = function(obj, iteratee, context) {
-    var result = -Infinity, lastComputed = -Infinity,
-        value, computed;
-    if (iteratee == null && obj != null) {
-      obj = isArrayLike(obj) ? obj : _.values(obj);
-      for (var i = 0, length = obj.length; i < length; i++) {
-        value = obj[i];
-        if (value > result) {
-          result = value;
-        }
-      }
-    } else {
-      iteratee = cb(iteratee, context);
-      _.each(obj, function(value, index, list) {
-        computed = iteratee(value, index, list);
-        if (computed > lastComputed || computed === -Infinity && result === -Infinity) {
+  // Generator function to create the max and min functions
+  var createExtremumFinder = function(max) {
+    return function(obj, iteratee, context) {
+      var keys = !isArrayLike(obj) && _.keys(obj),
+          length = (keys || obj).length,
+          result = max ? -Infinity : Infinity,
+          lastComputed = result,
+          found = false;
+      iteratee = iteratee == null ? null : cb(iteratee, context);
+      for (var index = 0; index < length; index++) {
+        var currentKey = keys ? keys[index] : index;
+        var value = obj[currentKey];
+        var computed = iteratee ? iteratee(value, currentKey, obj) : value;
+        if ((max ? computed > lastComputed : computed < lastComputed) || (!found && computed === result)) {
+          found = true;
```

Comment:
Well, the iteratee can return anything. So, if the first iteration returns `undefined`, that won't be greater than `-Infinity`. But, the next iteration might be equal.